test: add integration test for application-inventory generation

jmt-lab opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Description of changes:**
Introduces an integration test that will build bottlerocket twice. First it will build bottlerocket with the latest released version of twoliter. Then it will build bottlerocket with the current commits build. Afterwards it does a comparison to make sure the application-inventory.json file is generated correctly and matches the released version.


**Testing done:**
Ran `make integ`


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

sam-berning commented on tests/integration-tests/Cargo.toml:
```diff
@@ -5,8 +5,18 @@ edition = "2021"
+duct.workspace = true
+flate2.workspace = true
 libc.workspace = true
+lzma-rs = "0.3"
+once_cell = "1.21"
```
nit: I'd prefer we keep all of the dependencies workspace dependencies

cbgbt replied:
`cargo-deny` is supposed to be forbidding this in CI, but I think it failed here because our `cargo-deny` call doesn't turn on tests? It's not really clear why it didn't flag this.

sam-berning commented on tests/integration-tests/src/appinventory.rs:
```diff
@@ -0,0 +1,204 @@
+async fn find_latest_version(repository: &str) -> String {
+    // We make a get request to fetch the github releases of twoliter
+    let url = Url::parse(&format!(
+        "https://api.github.com/repos/bottlerocket-os/{repository}/releases"
```
Can we just use https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#get-the-latest-release to get the latest? Should be fine as long as we don't mark release candidates in Twoliter as "latest"

cbgbt commented on tests/integration-tests/src/appinventory.rs:
```diff
@@ -0,0 +1,204 @@
+        .get(url)
+        .header("Accept", "application/vnd.github+json")
+        .header("X-GitHub-Api-Version", "2022-11-28")
+        .header("User-Agent", "curl/8.9.1") // Pretend to be curl
```
Is it problematic to use a more accurate `User-Agent`?

cbgbt commented on tests/integration-tests/src/appinventory.rs:
```diff
@@ -0,0 +1,204 @@
+    let client = reqwest::Client::new();
+    let response = client
+        .get(url)
+        .header("Accept", "application/vnd.github+json")
```
We should probably optionally authenticate here if we're running in CI so that we don't get throttled. Using `octocrab` would probably make this a lot easier (and also do the nice things like settings the right headers).

cbgbt commented on tests/integration-tests/src/appinventory.rs:
```diff
@@ -0,0 +1,204 @@
+    (tmp_dir, twoliter_bin)
+}
+
+#[derive(Deserialize, Eq, Debug)]
+#[serde(rename_all = "PascalCase")]
+struct ContentView {
```
Not a hard blocker, but I'd prefer to hoist this as `pub` into something that we could re-use to perform inventory generation in Rust.

cbgbt commented on tests/integration-tests/src/appinventory.rs:
```diff
@@ -0,0 +1,204 @@
+    content: Vec<ContentView>,
+}
+
+impl PartialEq for ContentView {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name
```
Can we remove this? It's the same as the `derive` implementation.

cbgbt commented on tests/integration-tests/src/appinventory.rs:
```diff
@@ -0,0 +1,204 @@
+        [
+            "build",
+            "variant",
+            "aws-ecs-2",
+            "--project-path",
+            project_path,
+        ],
+        cmd_env,
+    );
+    assert!(
+        output.status.success(),
+        "failed to run twoliter build variant"
+    );
+    // Now we should have the application inventory file
+    let aif = tmp_dir.path().join(format!(
+        "build/images/x86_64-aws-ecs-2/latest/application-inventory.json",
```
nits:
* Use a `CONSTANT` for the target variant
* I think `x86_64` is correct since you've mentioned in person that it's `twoliter`'s default if no arch is specified, but we should probably explicitly provide an ARCH and then check for it here. This would protect us against an architecture-specific test breakage if we modify the default arch selection behavior.

cbgbt commented on tests/integration-tests/src/twoliter_update.rs:
```diff
@@ -1,13 +1,15 @@
+use crate::twoliter_build::copy_project_to_temp_dir;
```
Thanks for these improvements to the integration tests! Do you mind putting them into a separate commit?

cbgbt reviewed (commented):
This is great, thanks for writing this! Just a few minor suggestions.

jmt-lab committed:
test: change integration test registry to use a random port

jmt-lab committed:
test: add integration test for application-inventory generation

Introduces an integration test that will build bottlerocket twice.
First it will build bottlerocket with the latest released version of twoliter.
Then it will build bottlerocket with the current commits build. Afterwards
it does a comparison to make sure the application-inventory.json file is generated
correctly and matches the released version.

cbgbt commented on tests/integration-tests/src/appinventory.rs:
```diff
@@ -0,0 +1,198 @@
+        .header("Accept", "application/vnd.github+json")
+        .header("X-GitHub-Api-Version", "2022-11-28")
+        // Pretend to be curl
+        .header("User-Agent", "curl/8.9.1");
```
We really ought to use something more accurate here, like `"twoliter-ci"` or similar.

cbgbt commented on tests/integration-tests/src/appinventory.rs:
```diff
@@ -0,0 +1,198 @@
+    let twoliter_bin = tmp_dir.path().join("twoliter");
+    let latest = find_latest_version("twoliter").await;
+    let arch = env::consts::ARCH;
+    let response = reqwest::get(format!("https://github.com/bottlerocket-os/twoliter/releases/download/{latest}/twoliter-{arch}-unknown-linux-musl.tar.xz")).await.expect("failed to download twoliter release archive");
```
We ought to set our user-agent for this request as well.

cbgbt reviewed (approved):
Nice, thanks!

We chatted offline and `octocrab` can't be added to our workspace due to an unfortunate issue having to do with `snafu`, the `backtrace` crate, `std::backtrace::Backtrace`, and feature unification.

Basically, `octocrab` has a hard dependency on `snafu` using `std::backtrace::Backtrace`; however, our workspace configures `snafu` to use the `backtrace` crate via `tough` and `parse-datetime`.

To fix this, we need to make the `backtrace` usage in `tough` configurable (or just move to `std`, which has been stable since, I believe, Rust 1.61).
